tests/utilities/test_auto_restart.py

```diff
@@ -625,3 +651,245 @@ def test_fast_forward_sampler_with_distributed_sampler_and_iterative_dataset():
     mp.spawn(
         _test_fast_forward_sampler_with_distributed_sampler_and_iterative_dataset, args=(worldsize, ), nprocs=worldsize
     )
+
+
+@mock.patch.dict(os.environ, {'PL_FAULT_TOLERANT_TRAINING': "1"})
+@RunIf(max_torch="1.6")
+def test_fault_tolerant_not_supported():
+    with pytest.raises(MisconfigurationException, match="Restart is only supported with torch >= 1.7.0."):
+        _fault_tolerant_enabled()
+
+
+@pytest.mark.skipif(torch.cuda.is_available(), reason="This test takes around 15 sec and should be skipped in Azure CI")
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@RunIf(min_torch="1.7.0")
+def test_combined_dataloader_state_dict_and_reload():
+    """
+    This test makes sure the CombinedLoader used in the condition of Lightning properly
+    capture its children DataLoader states.
+    """
+
+    def create_iterable_dataset(batch_size, num_workers, attr_name="iter_sampler"):
+        dataset = RangeIterableDataset(
+            range(50), num_workers=num_workers, batch_size=batch_size, is_in_workers=True, attr_name=attr_name
+        )
+        return CaptureIterableDataset(dataset)
+
+    def create_dataloader():
+        dataset = range(50)
+        num_workers = 2
+        batch_size = 8
+        sampler = FastForwardSampler(SequentialSampler(dataset))
+        sampler.setup(batch_size)
+
+        dataloader = DataLoader(
+            dataset,
+            sampler=sampler,
+            batch_size=batch_size,
+        )
+        dataloader.fast_forward_sampler = sampler
+
+        loader_dict = {
+            "a": [
+                DataLoader(create_iterable_dataset(3, num_workers), num_workers=num_workers, batch_size=3),
+                dataloader,
+            ],
+            "b": DataLoader(
+                create_iterable_dataset(2, num_workers=1, attr_name="custom_sampler"), num_workers=0, batch_size=2
+            )
+        }
+        apply_to_collection(loader_dict, DataLoader, Trainer._add_sampler_metadata_collate)
+        return CombinedLoader(loader_dict)
+
+    # Lightning will wrap the iterator within a prefect function as follow.
+    def prefetch_iterator(iterable: Iterable):
+        it = iter(iterable)
+
+        try:
+            # the iterator may be empty from the beginning
+            last = next(it)
+        except StopIteration:
+            return
+
+        for val in it:
+            # yield last and has next
+            yield last, False, it
+            last = val
+        # yield last, no longer has next
+        yield last, True, it
+
+    dataloader = create_dataloader()
```

Comment:
These tests seem very convoluted

Why do we need all this logic for the tests? Is this trying to mimic some of the behaviour in data loading?

Can the code be re-structured so it's easier to unit-test without having to re-write part of the logic we want to test in the test itself?

Reply:
Yes, this is mimicking the perfecting behaviour from Lightning.
Feel free to simplify if needed.

Reply:
Moved the helpers function out the test.